File: clients/java/src/test/java/io/volant/AclsTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** Create/Delete/ListAcls client tests against a fake native server. */
class AclsTest {
    private static AclBinding sample() {
        return new AclBinding("User:alice", 0, "events", 3, 1);
    }

    @Test
    void createOk() throws Exception {
        AclBinding entry = sample();
        try (AclServer srv = new AclServer(0, 0, 0, 1, Collections.singletonList(entry))) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.createAcls(Collections.singletonList(entry));
            }
            assertEquals(Collections.singletonList(entry), srv.create.get());
            assertEquals(Collections.singletonList(Codec.OP_CREATE_ACLS), srv.opcodes);
        }
    }

    @Test
    void deleteReturnsRemoved() throws Exception {
        AclBinding entry = sample();
        try (AclServer srv = new AclServer(0, 0, 0, 1, Collections.emptyList())) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                int n = c.deleteAcls(Collections.singletonList(entry));
                assertEquals(1, n);
            }
            assertEquals(Collections.singletonList(entry), srv.delete.get());
        }
    }

    @Test
    void listReturnsBindings() throws Exception {
        AclBinding entry = sample();
        try (AclServer srv = new AclServer(0, 0, 0, 0, Collections.singletonList(entry))) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                List<AclBinding> listed = c.listAcls();
                assertEquals(Collections.singletonList(entry), listed);
            }
            Codec.ListAclsRequest req = srv.listReq.get();
            assertEquals("", req.principal);
            assertEquals(255, req.resourceType);
            assertEquals("", req.resource);
        }
    }

    @Test
    void unauthorizedRaises() throws Exception {
        try (AclServer srv = new AclServer(23, 0, 0, 0, Collections.emptyList())) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex =
                        assertThrows(BrokerException.class, () -> c.createAcls(Collections.singletonList(sample())));
                assertEquals(23, ex.code);
                assertEquals("create_acls", ex.op);
            }
        }
    }

    private static final class AclServer implements AutoCloseable {
        final int port;
        final AtomicReference<List<AclBinding>> create = new AtomicReference<>();
        final AtomicReference<List<AclBinding>> delete = new AtomicReference<>();
        final AtomicReference<Codec.ListAclsRequest> listReq = new AtomicReference<>();
        final List<Integer> opcodes = new CopyOnWriteArrayList<>();
        private final int createError;
        private final int deleteError;
        private final int listError;
        private final int removed;
        private final List<AclBinding> entries;
        private final ServerSocket listen;
        private final Thread thread;

        AclServer(int createError, int deleteError, int listError, int removed, List<AclBinding> entries)
                throws IOException {
            this.createError = createError;
            this.deleteError = deleteError;
            this.listError = listError;
            this.removed = removed;
            this.entries = entries;
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            thread = new Thread(this::serve, "volant-acls");
            thread.setDaemon(true);
            thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    opcodes.add(d.frame.opcode);
                    byte[] payload;
                    int respOp;
                    if (d.frame.opcode == Codec.OP_CREATE_ACLS) {
                        Codec.CreateAclsRequest req = Codec.decodeCreateAclsRequest(d.frame.payload);
                        create.set(req.entries);
                        payload = Codec.encodeCreateAclsResponse(new Codec.CreateAclsResponse(createError));
                        respOp = Codec.OP_CREATE_ACLS_RESPONSE;
                    } else if (d.frame.opcode == Codec.OP_DELETE_ACLS) {
                        Codec.DeleteAclsRequest req = Codec.decodeDeleteAclsRequest(d.frame.payload);
                        delete.set(req.entries);
                        payload = Codec.encodeDeleteAclsResponse(
                                new Codec.DeleteAclsResponse(deleteError, removed));
                        respOp = Codec.OP_DELETE_ACLS_RESPONSE;
                    } else if (d.frame.opcode == Codec.OP_LIST_ACLS) {
                        listReq.set(Codec.decodeListAclsRequest(d.frame.payload));
                        payload = Codec.encodeListAclsResponse(new Codec.ListAclsResponse(listError, entries));
                        respOp = Codec.OP_LIST_ACLS_RESPONSE;
                    } else {
                        return;
                    }
                    out.write(Frame.encode(respOp, d.frame.correlationId, payload));
                    out.flush();
                }
            } catch (Exception ignored) {
                // test assertions cover the client side
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: clients/go/acls_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

func sampleAcl() codec.AclBinding {
	return codec.AclBinding{
		Principal:    "User:alice",
		ResourceType: 0,
		Resource:     "events",
		Operation:    3,
		Permission:   1,
	}
}

type aclAdminResult struct {
	opcodes    []uint16
	create     []codec.AclBinding
	delete     []codec.AclBinding
	listReq    codec.ListAclsRequest
	err        error
}

func serveAcls(t *testing.T, createErr, deleteErr, listErr uint16, removed uint32, entries []codec.AclBinding) (addr string, got *aclAdminResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &aclAdminResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			res.opcodes = append(res.opcodes, f.Opcode)
			var (
				payload []byte
				op      uint16
			)
			switch f.Opcode {
			case codec.OpCreateAcls:
				req, err := codec.DecodeCreateAclsRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.create = req.Entries
				payload, err = codec.EncodeCreateAclsResponse(codec.CreateAclsResponse{ErrorCode: createErr})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpCreateAclsResponse
			case codec.OpDeleteAcls:
				req, err := codec.DecodeDeleteAclsRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.delete = req.Entries
				payload, err = codec.EncodeDeleteAclsResponse(codec.DeleteAclsResponse{ErrorCode: deleteErr, Removed: removed})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpDeleteAclsResponse
			case codec.OpListAcls:
				req, err := codec.DecodeListAclsRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.listReq = req
				payload, err = codec.EncodeListAclsResponse(codec.ListAclsResponse{
					ErrorCode: listErr,
					Entries:   entries,
				})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpListAclsResponse
			default:
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			raw, err := frame.Encode(op, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestCreateAclsOk(t *testing.T) {
	entry := sampleAcl()
	addr, got, stop := serveAcls(t, 0, 0, 0, 1, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if err := c.CreateAcls([]codec.AclBinding{entry}); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if len(got.create) != 1 || got.create[0] != entry {
		t.Fatalf("create %#v", got.create)
	}
	if len(got.opcodes) != 1 || got.opcodes[0] != codec.OpCreateAcls {
		t.Fatalf("opcodes %#v", got.opcodes)
	}
}

func TestCreateAclEncodesOneEntry(t *testing.T) {
	entry := sampleAcl()
	addr, got, stop := serveAcls(t, 0, 0, 0, 1, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if err := c.CreateAcl(entry); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if len(got.create) != 1 || got.create[0] != entry {
		t.Fatalf("create %#v", got.create)
	}
	if len(got.opcodes) != 1 || got.opcodes[0] != codec.OpCreateAcls {
		t.Fatalf("opcodes %#v", got.opcodes)
	}
}

func TestDeleteAclsReturnsRemoved(t *testing.T) {
	entry := sampleAcl()
	addr, got, stop := serveAcls(t, 0, 0, 0, 1, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	n, err := c.DeleteAcls([]codec.AclBinding{entry})
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if n != 1 {
		t.Fatalf("removed %d", n)
	}
	if len(got.delete) != 1 || got.delete[0] != entry {
		t.Fatalf("delete %#v", got.delete)
	}
}

func TestDeleteAclEncodesOneEntry(t *testing.T) {
	entry := sampleAcl()
	addr, got, stop := serveAcls(t, 0, 0, 0, 1, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	n, err := c.DeleteAcl(entry)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if n != 1 {
		t.Fatalf("removed %d", n)
	}
	if len(got.delete) != 1 || got.delete[0] != entry {
		t.Fatalf("delete %#v", got.delete)
	}
}

func TestListAclsReturnsBindings(t *testing.T) {
	entry := sampleAcl()
	addr, got, stop := serveAcls(t, 0, 0, 0, 0, []codec.AclBinding{entry})
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	listed, err := c.ListAcls("", 255, "")
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if len(listed) != 1 || listed[0] != entry {
		t.Fatalf("listed %#v", listed)
	}
	if got.listReq.Principal != "" || got.listReq.ResourceType != 255 || got.listReq.Resource != "" {
		t.Fatalf("list req %+v", got.listReq)
	}
}

func TestListAclsAllEncodesEmptyFilters(t *testing.T) {
	entry := sampleAcl()
	addr, got, stop := serveAcls(t, 0, 0, 0, 0, []codec.AclBinding{entry})
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	listed, err := c.ListAclsAll()
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if len(listed) != 1 || listed[0] != entry {
		t.Fatalf("listed %#v", listed)
	}
	if got.listReq.Principal != "" || got.listReq.ResourceType != 255 || got.listReq.Resource != "" {
		t.Fatalf("list req %+v", got.listReq)
	}
}

func TestCreateAclsUnauthorizedRaises(t *testing.T) {
	addr, _, stop := serveAcls(t, 23, 0, 0, 0, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	err = c.CreateAcls([]codec.AclBinding{sampleAcl()})
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 23 || be.Op != "create_acls" {
		t.Fatalf("broker error %+v", be)
	}
}
